src/ApparentHorizons/SpherepackIterator.hpp

```diff
@@ -0,0 +1,116 @@
+///
+/// \details
+/// The internal SPHEREPACK ordering is not intuitive, so
+/// SpherepackIterator exists for the purpose of iterating over the
+/// array and determining which specific values of l and m correspond
+/// to each iteration.
+/// \example
+/// \snippet Test_SpherepackIterator.cpp spherepack_iterator_example
+class SpherepackIterator {
+ public:
+  /// 'a' and 'b' are what SPHEREPACK calls the two FORTRAN arrays.
+  /// 'a' holds real and 'b' holds the imaginary parts of the coefficients.
+  enum class CoefficientArray { a, b };
+  SpherepackIterator(size_t l_max_input, size_t m_max_input, size_t stride = 1);
+
+  size_t l_max() const noexcept { return l_max_; }
+  size_t m_max() const noexcept { return m_max_; }
+  size_t n_th() const noexcept { return n_th_; }
+  size_t n_ph() const noexcept { return n_ph_; }
+  size_t stride() const noexcept { return stride_; }
+
+  /// Size of the combined a/b array *in the SPHEREPACK dimensions*.
+  /// (so the actual size of the array is 'stride' times larger).
+  size_t spherepack_array_size() const noexcept {
+    return (l_max_ + 1) * (m_max_ + 1) * 2;
+  }
+
+  SpherepackIterator& operator++() noexcept {
+    ASSERT(current_compact_index_ != offset_into_spherepack_array.size(),
+           "Incrementing an invalid iterator: "
+               << current_compact_index_ << " "
+               << offset_into_spherepack_array.size());
+    ++current_compact_index_;
+    return *this;
+  }
+
+  explicit operator bool() const noexcept {
+    return current_compact_index_ < offset_into_spherepack_array.size();
+  }
+
+  /// Current index into combined a/b array storage. Used to index a
+  /// vector representing spectral coefficients stored in SPHEREPACK
+  /// format.
+  size_t operator()() const noexcept {
+    return offset_into_spherepack_array[current_compact_index_];
+  }
+
+  /// current values of l and m, and of which array (real or imaginary parts)
+  size_t l() const noexcept { return compact_l_[current_compact_index_]; }
+  size_t m() const noexcept { return compact_m_[current_compact_index_]; }
+  CoefficientArray coefficient_array() const noexcept {
+    return (current_compact_index_ < number_of_valid_entries_in_a_
+                ? CoefficientArray::a
+                : CoefficientArray::b);
+  }
+
+  /// Resets back to beginning value, returns *this.
+  SpherepackIterator& reset() noexcept;
+
+  /// Go to specific value of l, m, array.  Returns *this.
+  SpherepackIterator& set(size_t l_input, size_t m_input,
+                          CoefficientArray coefficient_array_input) noexcept;
+
+  /// Same as 'set' above, but assumes CoefficientArray is 'a' for m>=0 and
+  /// 'b' for m<0.  This is useful when converting between true
+  /// spherical harmonics (which allow negative values of m) and
+  /// SPHEREPACK coefficients (which have only positive values of m,
+  /// but two arrays for real and imaginary parts).
```

Comment:
Should there be a similar interface for getting `(l,m)` out of the iterator? One that gives you the "true" (l,m) and doesn't need `a/b`? I don't know if this is "missing" because it actually isn't useful or because we never bothered...

Reply:
We can add if needed.  (But not sure what you mean by 'true' l,m.)

Reply:
(I was trying to reflect the wording used in the comment, but it wasn't clear. I meant to propose: an interface giving the l,m with m allowed to be negative, as opposed to the SPHEREPACK l,m where m is positive and so the `a/b` information is also needed.)